File: chess.py
#!/usr/bin/env python3

# Knights will be modeled as 'W' and 'B' for their color
# Using enums would be cleaner, but hey

from copy import deepcopy, copy


PIECE_MAPPING = {"W": "♘", "B": "♞", " ": "□"}


def string_layout(layout):
    """
    Returns  an ASCII representation of a layout, handy for "hashing"
    (as our layouts are dictionaries and hence not hashable)
    and for printing.
    Layouts are indexed by their column and row number,
    starting on the bottom left corner
    ' ' if empty, 'W' if White knight and 'B' if Black knight"""
    printout = ""
    for row in range(3, -1, -1):
        for col in range(4):
            position = (row, col)
            if position in layout:
                slot = PIECE_MAPPING[layout[position]]
            else:
                slot = " "
            printout = printout + slot
        printout = printout + "\n"
    return printout


class Chessboard:
    # fmt: off
    START_LAYOUT = {
        (0, 0): 'B', (0, 1): ' ', (0, 2): 'B', (0, 3): ' ',
                     (1, 1): ' ', (1, 2): 'W', (1, 3): ' ',
                     (2, 1): ' ', (2, 2): ' ',
                     (3, 1): 'W'
    }
    GOAL_LAYOUT = {
        (0, 0): 'W', (0, 1): ' ', (0, 2): 'W', (0, 3): ' ',
                     (1, 1): ' ', (1, 2): 'B', (1, 3): ' ',
                     (2, 1): ' ', (2, 2): ' ',
                     (3, 1): 'B'
    }
    # fmt: on
    def __init__(self, start_layout=START_LAYOUT, goal=GOAL_LAYOUT):
        """A Chessboard is a layout, mechanisms to mutate its layout and
        a place to keep its history."""
        self.layout = copy(start_layout)
        self.goal_layout = copy(goal)
        self.history = [copy(self.layout)]

    @property
    def finished(self):
        """Property when the board has reached the desired layout"""
        return self.layout == self.goal_layout

    def find_pieces(self, piece):
        """Gets the list of pieces"""
        return set(
            [value for value in self.layout.keys() if self.layout[value] == piece]
        )

    def print_history(self):
        """Returns the board's full history"""
        a = ""
        for the_layout in self.history:
            a = f"{a}\n{string_layout(the_layout)}"
        return a

    def move_piece(self, start_pos, end_pos):
        """Moves a piece while keeping a history of all moves.
        Returns True in case of success, False in case of failure.
        """
        if start_pos not in self.layout or end_pos not in self.layout:
            return False
        if self.layout[end_pos] != " ":
            return False
        if self.layout[start_pos] == " ":
            return False
        piece = self.layout[start_pos]
        self.layout[start_pos] = " "
        self.layout[end_pos] = piece
        self.history.append(deepcopy(self.layout))
        return True

    def next_boards(self):
        """Constructor for the possible next boards"""
        boards = []
        for start_pos in self.find_pieces("B").union(self.find_pieces("W")):
            # fmt: off
            for move in [(2, -1), (2, 1), (1, -2), (1, 2), (-2, -1), (-2, 1), (-1, 2), (-1, -2)]:
                # fmt: on
                new_board = deepcopy(self)
                end_pos = (start_pos[0]+move[0], start_pos[1]+move[1])
                if new_board.move_piece(start_pos, end_pos):
                    boards.append(new_board)
        return boards


def traverse(board):
    """Traverses all possible boards and returns the boards that reach the goal layout
    and the layouts it visited along the way (excluding the final layout...)"""
    finished_boards = []
    boards = [board]
    visited_layouts = {}
    while boards:
        visited_layouts.update({string_layout(b.layout): b.layout for b in boards})
        next_boards = [
            inner_board
            for outer_board in boards
            for inner_board in outer_board.next_boards()
            # This takes care of removing both
            # - looping boards (they are not interesting if they are not
            # finished by now)
            # - layouts that have already been visited
            if not inner_board.layout in visited_layouts.values()
        ]
        # We could decide to stop at the first finished board, but let's find more
        # (all?) of them instead
        finished_boards.extend(
            [the_board for the_board in next_boards if the_board.finished]
        )
        # We keep only one board per current layout
        # (so technically we throw the others' history away,
        # which may ignore finished boards - to be pondered)
        # and exclude all those that have reached the goal
        boards = {
            string_layout(the_board.layout): the_board
            for the_board in next_boards
            if not the_board.finished
        }.values()

    return finished_boards, visited_layouts.values()


def main():
    board = Chessboard()
    completed_boards, visited_layouts = traverse(board)
    for board in completed_boards:
        print("*********************")
        print(
            f"In {len(board.history)} moves, "
            f"visiting {len(visited_layouts)} layouts (excluding the final one):"
        )
        print(board.print_history())
        print()


if __name__ == "__main__":
    main()

File: chess.go
package main

import (
	"fmt"
	"strings"
)

// Position represents a (row, col) coordinate on the board
type Position struct {
	row, col int
}

// Layout is the board state: a map from Position to piece ('W', 'B', or ' ')
type Layout map[Position]rune

// pieceMappings maps piece chars to display symbols
var pieceMapping = map[rune]string{
	'W': "♘",
	'B': "♞",
	' ': "□",
}

// knightMoves lists all possible knight move offsets
var knightMoves = []Position{
	{2, -1}, {2, 1}, {1, -2}, {1, 2},
	{-2, -1}, {-2, 1}, {-1, 2}, {-1, -2},
}

var startLayout = Layout{
	{0, 0}: 'B', {0, 1}: ' ', {0, 2}: 'B', {0, 3}: ' ',
	{1, 1}: ' ', {1, 2}: 'W', {1, 3}: ' ',
	{2, 1}: ' ', {2, 2}: ' ',
	{3, 1}: 'W',
}

var goalLayout = Layout{
	{0, 0}: 'W', {0, 1}: ' ', {0, 2}: 'W', {0, 3}: ' ',
	{1, 1}: ' ', {1, 2}: 'B', {1, 3}: ' ',
	{2, 1}: ' ', {2, 2}: ' ',
	{3, 1}: 'B',
}

// stringLayout returns an ASCII representation of a layout, used for display and as a map key
func stringLayout(layout Layout) string {
	var sb strings.Builder
	for row := 3; row >= 0; row-- {
		for col := 0; col < 4; col++ {
			pos := Position{row, col}
			if piece, ok := layout[pos]; ok {
				sb.WriteString(pieceMapping[piece])
			} else {
				sb.WriteString(" ")
			}
		}
		sb.WriteString("\n")
	}
	return sb.String()
}

// copyLayout performs a deep copy of a layout
func copyLayout(src Layout) Layout {
	dst := make(Layout, len(src))
	for k, v := range src {
		dst[k] = v
	}
	return dst
}

// layoutsEqual returns true if two layouts are identical
func layoutsEqual(a, b Layout) bool {
	if len(a) != len(b) {
		return false
	}
	for k, v := range a {
		if b[k] != v {
			return false
		}
	}
	return true
}

// Chessboard holds the current layout, goal, and move history
type Chessboard struct {
	layout     Layout
	goalLayout Layout
	history    []Layout
}

// newChessboard creates a new board with the given start and goal layouts
func newChessboard(start, goal Layout) *Chessboard {
	layout := copyLayout(start)
	return &Chessboard{
		layout:     layout,
		goalLayout: goal,
		history:    []Layout{copyLayout(layout)},
	}
}

// finished returns true when the board has reached the goal layout
func (cb *Chessboard) finished() bool {
	return layoutsEqual(cb.layout, cb.goalLayout)
}

// findPieces returns all positions containing the given piece
func (cb *Chessboard) findPieces(piece rune) []Position {
	var positions []Position
	for pos, p := range cb.layout {
		if p == piece {
			positions = append(positions, pos)
		}
	}
	return positions
}

// printHistory returns a string of the full board history
func (cb *Chessboard) printHistory() string {
	var sb strings.Builder
	for _, layout := range cb.history {
		sb.WriteString("\n")
		sb.WriteString(stringLayout(layout))
	}
	return sb.String()
}

// movePiece moves a piece from start to end, recording the move in history.
// Returns true on success, false on failure.
func (cb *Chessboard) movePiece(start, end Position) bool {
	if _, ok := cb.layout[start]; !ok {
		return false
	}
	if _, ok := cb.layout[end]; !ok {
		return false
	}
	if cb.layout[end] != ' ' {
		return false
	}
	if cb.layout[start] == ' ' {
		return false
	}
	piece := cb.layout[start]
	cb.layout[start] = ' '
	cb.layout[end] = piece
	cb.history = append(cb.history, copyLayout(cb.layout))
	return true
}

// deepCopy returns a full deep copy of the chessboard
func (cb *Chessboard) deepCopy() *Chessboard {
	historyCopy := make([]Layout, len(cb.history))
	for i, h := range cb.history {
		historyCopy[i] = copyLayout(h)
	}
	return &Chessboard{
		layout:     copyLayout(cb.layout),
		goalLayout: copyLayout(cb.goalLayout),
		history:    historyCopy,
	}
}

// nextBoards generates all valid next board states from the current one
func (cb *Chessboard) nextBoards() []*Chessboard {
	var boards []*Chessboard
	var allPieces []Position
	allPieces = append(allPieces, cb.findPieces('B')...)
	allPieces = append(allPieces, cb.findPieces('W')...)

	for _, startPos := range allPieces {
		for _, move := range knightMoves {
			endPos := Position{startPos.row + move.row, startPos.col + move.col}
			newBoard := cb.deepCopy()
			if newBoard.movePiece(startPos, endPos) {
				boards = append(boards, newBoard)
			}
		}
	}
	return boards
}

// traverse performs BFS over all possible board states, returning boards that
// reach the goal and all visited layouts along the way
func traverse(board *Chessboard) ([]*Chessboard, []Layout) {
	var finishedBoards []*Chessboard
	boards := []*Chessboard{board}
	visitedLayouts := make(map[string]Layout)

	for len(boards) > 0 {
		// Record all current layouts as visited
		for _, b := range boards {
			key := stringLayout(b.layout)
			visitedLayouts[key] = copyLayout(b.layout)
		}

		// Generate next boards, excluding already-visited layouts
		uniqueNext := make(map[string]*Chessboard)
		for _, outerBoard := range boards {
			for _, innerBoard := range outerBoard.nextBoards() {
				key := stringLayout(innerBoard.layout)
				if _, visited := visitedLayouts[key]; visited {
					continue
				}
				if innerBoard.finished() {
					finishedBoards = append(finishedBoards, innerBoard)
				} else {
					// Keep only one board per layout (discard duplicate histories)
					uniqueNext[key] = innerBoard
				}
			}
		}

		boards = make([]*Chessboard, 0, len(uniqueNext))
		for _, b := range uniqueNext {
			boards = append(boards, b)
		}
	}

	visited := make([]Layout, 0, len(visitedLayouts))
	for _, v := range visitedLayouts {
		visited = append(visited, v)
	}
	return finishedBoards, visited
}

func main() {
	board := newChessboard(startLayout, goalLayout)
	completedBoards, visitedLayouts := traverse(board)
	for _, b := range completedBoards {
		fmt.Println("*********************")
		fmt.Printf("In %d moves, visiting %d layouts (excluding the final one):\n",
			len(b.history), len(visitedLayouts))
		fmt.Println(b.printHistory())
		fmt.Println()
	}
}
